lib/vdsm/virt/thinp.py

```diff
@@ -262,67 +303,76 @@ def _extend_drive_if_needed(self, drive):
         Check if a drive should be extended, and start extension flow if
         needed.
         """
-        if drive.threshold_state == storage.BLOCK_THRESHOLD.SET:
-            # Should never happen; we monitor only UNSET and EXCEEDED drives.
+        if drive.threshold_state == storage.BLOCK_THRESHOLD.UNSET:
+            self._handle_unset(drive)
+        elif drive.threshold_state == storage.BLOCK_THRESHOLD.SET:
+            # Extend completed during monitoring, nothing to do.
+            pass
+        elif drive.threshold_state == storage.BLOCK_THRESHOLD.EXCEEDED:
+            self._handle_exceeded(drive)
+        else:
             self._log.warning(
-                "Unexpected state for drive %s: threshold_state SET",
-                drive.name)
-            return
+                "Unexpected threshold state %s for drive %s",
+                drive.threshold_state, drive.name)
 
-        block_info = drive.block_info
+    def _handle_unset(self, drive):
+        """
+        Handle drive with UNSET threshold state.
+        """
+        if self._drive_needs_extend(drive):
+            # The threshold was already exceeded, so qemu will never trigger a
+            # block threshold event. Synthesize a block threshold event, and
+            # handle the drive as if we received an event right now.
+            drive.on_block_threshold(drive.path)
+            self._handle_exceeded(drive, urgent=True)
+        else:
+            # Set block threshold. on_block_threshold() will be called when the
+            # guest exceeds the threshold. If we will miss the event,
+            # on_enospc() will be called.
+            self._set_threshold(
+                drive, drive.block_info.physical, drive.block_info.index)
 
-        # Set block threshold if needed. However since this is racy, and we may
-        # not get a block threshold event, continue and check if the volume
-        # needs extension.
-        if drive.threshold_state == storage.BLOCK_THRESHOLD.UNSET:
-            self._set_threshold(drive, block_info.physical, block_info.index)
+    def _handle_exceeded(self, drive, urgent=False):
+        """
+        Handle drive with EXCEEDED threshold state.
+        """
+        if not self._can_extend_drive(drive):
+            # Can happen drive size < 25g, drive is extended to maximum size,
```

Comment:
... happen *if* drive ...

Is 25g some magic value or is another value missing in the description?

Reply:
It depends on the chunk size and utilization. When do the last extend
to capacity bytes, free space is set to capacity - 2g.

For example 20g image:

    [            20g      ] 2g ]
                          ^
                         20g

Because we allow 1.1 * 20g (22g), the block threshold is now 20g.

Because writing 20g of data needs more than 20g (for qcow2 metadata)
the guest can write after offset 20g.

With bigger images (30g) we don't see this issue. The next extend
is complete chunk (2.5g):

    [            30g              ]  2.5g ]
                                     ^
                                   30.5g

And the threshold is at 30.5g. Practically the guest never exceeds 
30.5, since qcow2 metadata is less than 1%.

I'm not sure what is the exact limit when this can happen, but 25g is good
estimate. I will change the comment to make this more clear.

We have a bug to reduce the maximum allocation since allocating extra 10%
is way too much and never needed.

Reply:
This cannot happen with 23g volume:

    maximum write offset: 23 * 1.01 -> 23.23
    maximum allocation: 23 * 1.1 -> 25.30
    threshold: 23.30

But it can happen with 22g volume:

    maximum write offset: 22 * 1.01 -> 22.22
    maximum allocation: 22 * 1.1 -> 24.20
    threshold: 22.20

Based 1% extra allocation for qcow2 metadata. Actual value is usually smaller.